src/quick-lint-js/fe/parse.cpp

```diff
@@ -734,6 +734,55 @@ void parser::consume_semicolon() {
   }
 }
 
+void parser::error_on_pointless_nullish_coalescing_operator(
+    expression::binary_operator* ast) {
+  
+  auto is_nullish_operator = [](string8_view s) -> bool {
+    return s == u8"??"_sv;
+  };
+
+  auto no_left_operand = [](string8_view s) -> bool {
+    return s == u8"??"_sv;
+  };
+
+  if (no_left_operand(ast->span().string_view())){
+    return;
+  }
+
+  for (span_size i = 0; i < ast->child_count() - 1; i++) {
+    if (no_left_operand(ast->child(0)->span().string_view())){
+        return;
+    }
+    source_code_span op_span = ast->operator_spans_[i];
+    string8_view s = op_span.string_view();
+    if (is_nullish_operator(op_span.string_view())) {
+        if (i >= 1){
+            // lhs is a multi-child expression
+            this->diag_reporter_->report(diag_pointless_nullish_coalescing_operator{
+          .question_question = op_span});
+        } else {
+            this->check_lhs_for_null_potential(ast->child(i),
+                                               ast->child(i + 1), op_span);
+        }
+    }
+  }
+}
+
+void parser::check_lhs_for_null_potential(expression* lhs, expression* rhs,
+                                           source_code_span op_span) {
+  switch(lhs->kind()){
+    case expression_kind::variable:
+        return;
+    case expression_kind::call:
+        return;
+    default:
+        // expressions that are not variables or function calls are never null
```

Comment:
Must fix: This is incorrect. There are other kinds of expressions, such as property lookups and parenthesized expressions, which might return null or undefined. The following uses of `??` all yield `true` but your code incorrectly issues warnings about them:
```javascript
async function f() {
  "use strict";

  console.log(null ?? true);
  var foo = null;
  console.log((foo) ?? true);
  console.log({}.missingProp ?? true);
  console.log({}['missingProp'] ?? true);
  console.log(await foo ?? true);
  console.log(void 42 ?? true);
  function bar() { return null; }
  console.log(bar`hello` ?? true);
  console.log(this ?? true);
}
f();
```
I think you should use a *whitelist* of never-null/undefined expressions rather than a *blacklist* of sometimes-null/undefined expressions.

Reply:
This was my approach at first, but I switched to blacklist since I could not figure out how to deal with multichild lhs expression when there were no parentheses. However, when trying to find examples to go off from, i found that other diagnostics like error_on_pointless_string_compare seem to assume that the left and right operands will be wrapped in parentheses, so i'm not sure its necessary for me to account for that.